flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/PojoSerializer.java

```diff
@@ -163,25 +164,42 @@ public boolean isImmutableType() {
 
     @Override
     public PojoSerializer<T> duplicate() {
+        TypeSerializer<Object>[] duplicateFieldSerializers = duplicateSerializers(fieldSerializers);
+        TypeSerializer<Object>[] duplicateRegisteredSerializers =
+                duplicateSerializers(registeredSerializers);
+
+        // we must create a new instance, otherwise the subclassSerializerCache can create
+        // concurrency problems
+        return new PojoSerializer<>(
```

Comment:
I am still having a hard time understanding the fix. Conceptually it makes sense but how does the other ctor interact with the snapshot? To me, it looks like the same fields are passed to the POJO serializer.

Reply:
That's the issue that the other ctor is not used when recreating the serializer from the snapshot. The other ctor discards parts of the internal state that potentially comes from the snapshot. Therefore it cannot be used for duplication. It can be used though for creating a serializer for the first time.

Reply:
BTW, the fields are not the same in both ctors. In the "old" ctor, we create
```
            LinkedHashMap<Class<?>, Integer> registeredClasses,
            TypeSerializer<?>[] registeredSerializers,
            Map<Class<?>, TypeSerializer<?>> subclassSerializerCache,
```
from the `ExecutionConfig` instead of retrieving them from the snapshot.

Reply:
Ah thanks now it make sense 🙏